enhancements/network/ovs-hardware-offload.md

```diff
@@ -0,0 +1,242 @@
+---
+title: ovs-hardware-offload
+authors:
+  - "@zshi"
+reviewers:
+  - "@dcbw"
+  - "@squeed"
+  - "@danwinship"
+  - "@pliurh"
+  - "@dougbtv"
+  - "@s1061123"
+approvers:
+  - "@knobunc"
+  - "@fepan"
+creation-date: 2020-04-27
+last-updated: 2020-04-27
+status: implementable
+see-also: []
+replaces: []
+superseded-by: []
+
+---
+
+# OVS-Hardware-Offload
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [X] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+- When will RHCOS uses RHEL-8.3 as base image?
+- Any known issues with connction track feature using OVS Hardware Offload?
+- Any known issues with IPv6 traffic using OVS Hardware Offload?
+
+## Summary
+
+This enhancement is to introduce OVS hardware offload feature in OpenShift
+for high-performance datapath with OVN-Kubernetes. OVS hardware offload
+enables offloading of OVS datapath rules to a compatible network card using
+linux Traffic Control (TC) tool `tc`. It relies on SR-IOV technology to control
+the SR-IOV Virtual Functions (VF) via VF representor port that is plugged
+into OpenvSwitch bridge. It improves the utilization of available bandwidth
+on underlying NICs by moving CPU intensive network packet processing from
+OVS software to hardware.
+
+## Motivation
+
+Enable high-performance datapath on primary pod interface with production-ready
+OVN-Kubernetes.
+
+### Goals
+
+Enable OpenFlow hardware offload for OVN datapath
+
+### Non-Goals
+
+Replace OpenFlow software based solution
+
+## Proposal
+
+OVS Hardware Offload accelerates the OVS software flows with a SR-IOV NIC
+that works in `switchdev` mode. It requires configurations in both OpenvSwitch
+daemon and SR-IOV hardware NIC.
+
+- OpenvSwitch daemon is managed by Cluster Network Operator (CNO) and runs in
+`ovs-node` DaemonSet. A boolean configurable field will be added in CNO
+`Network` Custom Resource Definition `spec.defaultNetwork.ovnKubernetesConfig`
+to enable or disable OVS Hardware Offload feature for OpenvSwitch daemon.
+
+- SR-IOV hardware NIC is managed by SR-IOV Network Operator and will be
+provisioned with a new device type `switchdev`. The initial supported NIC
+model will be Mellanox MT27800 Family [ConnectX-5] 25GbE dual-port SFP28.
+
+Besides CNO and SR-IOV Operator changes, there are several other dependencies.
+
+- OVS Hardware Offload will only be supported on bare metal deployment.
+
+- Multus supports overwriting cluster network definition by specifying a
+customized default net-attach-def in pod annotation. Customized net-attach-def
+will still use `ovn-k8s-cni-overlay` CNI plugin, but with an special annotation
+that requests a SR-IOV device.
```

Comment:
There needs to be a place where user can request a hw offload pod/interface, currently the design relies on user to specify default cluster-network annoatation, how would admission controller knows whether user wants to have a hw-offload pod?

Reply:
We can use a pod annotation to indicate whether a pod needs HW offloading.

Reply:
Re the wording, I'll update and add the owner of actions to make it clear who will do what.

Current pod creation flow is the same as how normal SR-IOV pod is created, user first defines a SR-IOV net-attach-def via SR-IOV Operator API, then specifies the net-attach-def in pod annotation. If we were to add a another way to declare the use of ovs hw offload pod, it would be inconsistent.  The customized net-attach-def for ovs hw offload in my opinion is not complex as below:

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: switch
  annotations:
    k8s.v1.cni.cncf.io/resourceName: openshift.io/switchdev_network
spec:
  Config: '{"cniVersion":"0.3.1","name":"ovn-kubernetes","type":"ovn-k8s-cni-overlay","ipam":{},"dns":{}}'
```

It has the same `spec.Config` body as the default net-attach-def in Multus CNI config, the only difference is the annotation part which varies per resource name (randomly defined by cluster network admin). so it will not be a fixed net-attach-def that can be pre-defined, but I agree that with the help of admission controller (whether it's multus or sriov), it can be achieved by adding a different annotation in pod to declare the use of ovs hw offload and the annotation should contain the desired sr-iov resource name, then admission controller would inspect each pod annotation, create and add the customized net-attach-def for ovs-hw-offload pod.

Re security / authorization,  it will have the same security /authorization as was applied to normal net-attach-def, like namespace isolation etc.  An extra security check may need to be added for the customized net-attach-def to make sure the body of net-attach-def is not configured with any unexpected values in ipam/dns.

Reply:
Updated to create customized net-attach-def automatically by sriov-operator which knows the switchdev resource name to be used in the net-attach-def.